src/base.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+#
+# SPDX-License-Identifier: LGPL-2.1-or-later
+#
+# Copyright (C) 2022 Collabora Limited
+# Author: Guillaume Tucker <guillaume.tucker@collabora.com>
+
+import logging
+import os
+import traceback
+
+import kernelci
+import kernelci.db
+
+from logger import Logger
+
+
+class Service:
+    """Common class for pipeline services"""
+
+    class Context:
+        """Generic object to handle arbitrary context data"""
+        def __init__(self, data=None, sub_id=None):
+            self._data = data or {}
+            self._sub_id = sub_id
+
+        @property
+        def data(self):
+            return self._data
+
+        @property
+        def sub_id(self):
+            return self._sub_id
```

Comment:
How about adding `__gettattr__` to access attributes from `self._data` dictionary with a dotted notation e.g. 
```
context = Context(data={'key': 'value'})
context.key

```


```suggestion
    class Context:
        """Generic object to handle arbitrary context data"""
        def __init__(self, data=None, sub_id=None):
            self._data = data or {}
            self._sub_id = sub_id

        @property
        def data(self):
            return self._data

        @property
        def sub_id(self):
            return self._sub_id

        def __getattr__(self, name):
            return self._data.get(name)
```

Reply:
What's the point in doing that?

It means setting a value using `ctx.data[key] = value` and getting it either the same way or with `key.value`.  So I think it's a bit more complicated than `Context(object)` and I don't see the advantage.  Having just a dictionary attribute makes it entirely explicit and also avoids clashes with other attributes e.g. `.sub_id`.  So I know there are many ways to solve this, but what's the reason for choosing one over another?

Reply:
I guess `__setattr__()` implementation can always be added to avoid `ctx.data[key] = value` ;)

The `Context` object is some kind of an environment data container for a `Service` class.  I don't exactly understand why there is a need of explicitly having some of it being accessed differently than `sub_id`. Having said that I understand that `sub_id` is something which will be present in the majority of services and we want to have it added explicitly to the `Context`. So I thought `__gettatr__` could unify a way to access `sub_id` and other data from context.

Apart from that you can add logic to the `__getatr__` that can reduce the usage of 
```
 key = ctx.data 'key', None)
 if key:
      do_something(key)
```
pattern in the `_stop` method of `Service` derived classes.

It also easier to refactor the code which has

```python
ctx.key = value
```

then 

```python
ctx.data[key] = value

# or

if ctx.data.get(key):
   do_something()
```
as the dotted notation is less likely to confuse IDEs "Find usages..." or "Redactor..." features etc.

The `__gettatr__` is called only when the attribute wasn't found by `__getattribute__`, so if `sub_id` is present in the class it won't be looked for in the `Context.data`.

Having said all of that, as you have mentioned, there are many ways to solve this. I find the dotted notation less complex than retrieving data form the `ctx.data` dictionary, because of the reasons above. Maybe I'm also a bit inspired by the `Meta` internal class of the Django ORM models.